File: Appointment/AppointmentTableViewCell.m
//
//  AppointmentTableViewCell.m
//  Appointment
//
//  Created by Jeffrey Moran on 6/9/17.
//  Copyright © 2017 Jeff Moran. All rights reserved.
//

#import "AppointmentTableViewCell.h"

@implementation AppointmentTableViewCell

- (void)awakeFromNib {
	[super awakeFromNib];
	// Initialization code
}

- (void)setSelected:(BOOL)selected animated:(BOOL)animated {
	[super setSelected:selected animated:animated];
	
	// Configure the view for the selected state
}

- (void)setAppointment:(Appointment *)anAppointment {
	appointment = anAppointment;
	
	[self setUpAppointmentValues];
}

- (instancetype)initWithStyle:(UITableViewCellStyle)style reuseIdentifier:(NSString *)reuseIdentifier {
	self = [super initWithStyle:style reuseIdentifier:reuseIdentifier];
	
	if (self) {
		[self addSubviews];
		[self setUpConstraints];
	}
	
	return self;
}

- (void)addSubviews {
	self.accessoryType = UITableViewCellAccessoryDisclosureIndicator;
	
	appointmentClientNameLabel = [[UILabel alloc] init];
	appointmentClientNameLabel.translatesAutoresizingMaskIntoConstraints = NO;
	
	appointmentTimeLabel = [[UILabel alloc] init];
	appointmentTimeLabel.translatesAutoresizingMaskIntoConstraints = NO;
	
	appointmentAddressLabel = [[UILabel alloc] init];
	appointmentAddressLabel.translatesAutoresizingMaskIntoConstraints = NO;
	appointmentAddressLabel.textColor = [UIColor darkGrayColor];
	appointmentAddressLabel.font = [UIFont systemFontOfSize:15];
	
	[self.contentView addSubview:appointmentClientNameLabel];
	[self.contentView addSubview:appointmentAddressLabel];
	[self.contentView addSubview:appointmentTimeLabel];
}

- (void)setUpConstraints {
	[NSLayoutConstraint
	 activateConstraints:@[
						   [appointmentClientNameLabel.leftAnchor constraintEqualToAnchor:self.contentView.leftAnchor constant: 15],
						   [appointmentClientNameLabel.topAnchor constraintEqualToAnchor:self.contentView.topAnchor constant: 5],
						   [appointmentClientNameLabel.rightAnchor constraintEqualToAnchor:self.contentView.rightAnchor constant: -15],
						   
						   [appointmentTimeLabel.leftAnchor constraintEqualToAnchor:appointmentClientNameLabel.leftAnchor],
						   [appointmentTimeLabel.topAnchor constraintEqualToAnchor:appointmentClientNameLabel.bottomAnchor constant: 5],
						   [appointmentTimeLabel.rightAnchor constraintEqualToAnchor:appointmentClientNameLabel.rightAnchor],
						   
						   [appointmentAddressLabel.leftAnchor constraintEqualToAnchor:appointmentClientNameLabel.leftAnchor],
						   [appointmentAddressLabel.topAnchor constraintEqualToAnchor:appointmentTimeLabel.bottomAnchor constant: 5],
						   [appointmentAddressLabel.rightAnchor constraintEqualToAnchor:appointmentClientNameLabel.rightAnchor],
						   [appointmentAddressLabel.bottomAnchor constraintEqualToAnchor:self.contentView.bottomAnchor constant: -5],
						   ]];
}

- (void)setUpAppointmentValues {
	appointmentClientNameLabel.text = !([appointment.clientName isEqualToString:@""]) ? appointment.clientName : @"Client name unavailable";
	
	appointmentAddressLabel.text = !([appointment.address isEqualToString:@""]) ? [NSString stringWithFormat:@"%@ %@", appointment.address, appointment.zipCode] : @"Property address unavailable";
	
	appointmentTimeLabel.text = appointment.appointmentTime ? appointment.appointmentDateString : @"Appointment time unavailable";
	
	self.backgroundColor = appointment.appointmentTime.timeIntervalSinceNow < 0.0 ? [UIColor colorWithRed:255.0 / 255.0 green:0.0 blue:0.0 alpha:0.1] : [UIColor whiteColor];
}

@end

File: Appointment/AppointmentTableViewCell.swift
//
//  AppointmentTableViewCell.swift
//  Appointment
//
//  Created by Jeffrey Moran on 10/14/20.
//  Copyright © 2020 Jeff Moran. All rights reserved.
//

import UIKit

class AppointmentTableViewCell: UITableViewCell {

    // MARK: - Static Constants

    static let reuseIdentifier = "appointmentTableViewCellIdentifier"

    // MARK: - Private Properties

    private lazy var stackView: UIStackView = {
        let stackView = UIStackView()
        stackView.translatesAutoresizingMaskIntoConstraints = false
        stackView.axis = .vertical
        stackView.spacing = 5.0

        return stackView
    }()

    private lazy var appointmentClientNameLabel: UILabel = {
        let label = UILabel()
        label.textColor = .label

        return label
    }()

    private lazy var appointmentTimeLabel: UILabel = {
        let label = UILabel()
        label.textColor = .secondaryLabel

        return label
    }()

    private lazy var appointmentAddressLabel: UILabel = {
        let label = UILabel()
        label.textColor = .tertiaryLabel

        return label
    }()

    // MARK: - Initializers

    override init(style: UITableViewCell.CellStyle, reuseIdentifier: String?) {
        super.init(style: style, reuseIdentifier: reuseIdentifier)

        accessoryType = .disclosureIndicator

        addSubviews()
        setUpConstraints()
    }

    @available(*, unavailable)
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    // MARK: - Private Methods

    private func addSubviews() {
        contentView.addSubview(stackView)
        stackView.addArrangedSubview(appointmentClientNameLabel)
        stackView.addArrangedSubview(appointmentTimeLabel)
        stackView.addArrangedSubview(appointmentAddressLabel)
    }

    private func setUpConstraints() {
        NSLayoutConstraint.activate([
            stackView.leadingAnchor.constraint(equalTo: contentView.readableContentGuide.leadingAnchor),
            stackView.trailingAnchor.constraint(equalTo: contentView.readableContentGuide.trailingAnchor),
            stackView.topAnchor.constraint(equalTo: contentView.readableContentGuide.topAnchor),
            stackView.bottomAnchor.constraint(equalTo: contentView.readableContentGuide.bottomAnchor)

        ])
    }

    // MARK: - Internal

    func style(with appointment: Appointment) {
        #warning("Rewritten from Obj-C -- clean this up")

        appointmentClientNameLabel.text = !appointment.clientName.isEmpty ? appointment.clientName : "Client name unavailable"

        let address = "\(appointment.address)\(appointment.zipCode)"
        appointmentAddressLabel.text = !address.isEmpty ? address : "Property address unavailable"

        appointmentTimeLabel.text = appointment.appointmentDateString
    }
}
